### ISSUE TITLE
pinning cosa on releases

### ISSUE THREAD

**Issue by dustymabe** [coreos/fedora-coreos-streams#64]
Right now we have a testing stream and we have a stable stream. We let things bake in testing a few weeks before promoting to stable, except for all of the changes that have gone into COSA, which could be significant. Maybe we should add a tag to cosa on the day a testing release is run and use that same cosa image that was used for testing when we do the stable pipeline run.

**Comment by jlebon** [comment]
Yeah, I think matching the cosa image makes sense. Hmm, I think we could have this work transparently by just using the image digest. E.g. when building `stable`, figure out from which `testing` release it's been promoted, then use the same cosa image digest that was used to build it.

Though this is dependent on the GC policy of Quay.io. If we build cosa in OpenShift, I think we might have more control.

**Comment by dustymabe** [comment]
> Though this is dependent on the GC policy of Quay.io. If we build cosa in OpenShift, I think we might have more control.

I think we'd just need to add a tag in quay.io whenever we do a testing build

**Comment by miabbott** [comment]
There was an ad-hoc discussion yesterday where this topic came up; it was suggested we could experiment for a couple of releases with doing a `git tag` of cosa on each `testing` release and see if we end up using the same tag to do the `stable` release.

**Comment by jlebon** [comment]
Another thing we could do is make the cosa tag to use live in https://github.com/coreos/fedora-coreos-config and have `bump-lockfile` also bump it.

Some advantages of this is that (1) cosa is "locked" by default which improves reproducibility, and (2) we transparently get cosa tag propagation as part of the regular promotion process. A disadvantage is that it introduces more delay between cosa changes and it affecting FCOS builds (though for hotfixes, we always have the override approach at job run time).

**Comment by bgilbert** [comment]
Adding context from https://github.com/coreos/coreos-assembler/issues/2644:

There's a second problem with the current situation: some cosa changes interlock with fedora-coreos-config changes, and sometimes the cosa side is accidentally merged before the f-c-c side reaches FCOS `stable`.  This breaks the build for some or all production branches.  The breakage might be fixed by routine branch promotion without anyone noticing, but it also impedes any out-of-cycle releases on existing branch heads that might be necessary before then.  In addition, if `testing` gets broken, the next `stable` promotion will be broken also.

Another fix approach is to use separate cosa branches for each FCOS stream, similar to how RHCOS branches cosa for each release.  Since the FCOS uses rolling branches, the cosa branches would also need to roll, using promotions similar to those in fedora-coreos-config.

In the shorter term, or if we don't want to address this in other ways, it would still be useful to catch FCOS branch breakage in cosa CI.  It'd probably be sufficient to run an additional CI job that performs a complete build and test cycle against the fedora-coreos-config `stable` branch.
